tools-local/unix/search-tool.sh

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env bash
+
+# Searches for the specified tool in the environment path, and the path within the repository as specified in the .toolversions file.
+
+usage()
+{
+    echo "Usage: $0 ToolName [-StrictToolVersionMatch]"
+    echo "  ToolName: Name of the tool to download."
+    echo "  (Optional) -StrictToolVersionMatch: If specified then, search will ensure that the version of the tool searched is the declared version."
+    echo "                                      Otherwise, search will attempt to find a version of the tool, which may not be the declared version."
+    echo ""
+    echo "Searches for the tool in the environment path, and the path within the repository as specified in the .toolversions file."
+    echo "If search is successful then, returns the path to the tool."
+    echo "Exit 1 if search fails to find the tool."
+}
+
+if [ $# -lt 1 ]; then
+    usage
+    exit 1
+fi
+
+if [ -z "$1" ]; then
+    echo "Argument passed as ToolName is empty. Please provide a non-empty string."
+    exit 1
+fi
+
+toolName="$1"
+shift
+strictToolVersionMatch=0
+
+while :; do
+    if [ $# -le 0 ]; then
+        break
+    fi
+
+    lowerI="$(echo $1 | awk '{print tolower($0)}')"
+    case $lowerI in
+        -stricttoolversionmatch)
+            strictToolVersionMatch=1
+            ;;
+        *)
+            usage
+            exit 1
+    esac
+    shift
+done
+
+scriptPath="$(cd "$(dirname "$0")"; pwd -P)"
+. "$scriptPath/tool-helper.sh"
+declaredVersion="$(get_tool_config_value "$toolName" "DeclaredVersion")"
+
+
+# Displays the tool path, and exits the script.
+display_tool_path()
+{
+    echo "$toolPath"
+    log_message "$toolName is available at $toolPath. Version is $toolVersion."
+}
+
+# Searches the tool in environment path.
+search_environment()
+{
+    log_message "Searching for $toolName in environment path."
+    hash "$toolName" 2>/dev/null
+
+    if [ $? -eq 0 ]; then
+        toolPath="$(which $toolName)"
+        toolVersion="$("$scriptPath/invoke-extension.sh" "get-version.sh" "$toolName" --ToolPath "$toolPath")"
+
+        if [ $strictToolVersionMatch -eq 0 ]; then
+            # No strictToolVersionMatch. Hence, return the path found without further checks.
+            display_tool_path
+            exit
+        else
+            # If strictToolVersionMatch is required then, ensure the version in environment path is same as declared version.
+            # If version matches then, return the path.
+            if [ "$toolVersion" == "$declaredVersion" ]; then
+                # Version available in environment path is the declared version.
+                display_tool_path
+                exit
+            fi
+        fi
```

Comment:
To copy over some [comments](https://github.com/ravimeda/corefx/pull/2/files#r109258953):

> 
> [dagood]
> I think this could be inverted and simplified:
> 
> ```sh
> if [ "$strictToolVersionMatch" -eq 1 ] && [ "$toolVersion" != "$declaredVersion" ]; then
>     log "Version of $toolName at..."
>     return 1
> fi
> echo "$toolPath"
> log "$toolName is available at $toolPath. Version is $toolVersion."
> ```

...

> [ravimeda]
> Consider the case where build needs the declared version 3.6.0. Say, the environment path CMake is 3.5.0. In that case as per your suggestion, return1, which I assume is same as exit 1 would terminate the search. Hence I suggest not to exit at this point, and let the script perform search_repository before exiting. Let me know if this clarifies...

> [dagood]
> Ah, it isn't the same. `return` will actually return from a function, setting error code.